### PR TITLE
Add opt-einsum to CI

### PR DESCRIPTION
**Pull request by janeyx99**
Depends on https://github.com/pytorch/pytorch/pull/84890.

This PR adds opt_einsum to CI, enabling path optimization for the multi-input case. It also updates the installation sites to install torch with einsum, but those are mostly to make sure it would work on the user's end (as opt-einsum would have already been installed in the docker or in prior set up steps).

This PR also updates the windows build_pytorch.bat script to use the same bdist_wheel and install commands as on Linux, replacing the `setup.py install` that'll become deprecated.

### CODE REVIEW

File: torch/functional.py [hunk 1]
```diff
@@ -15,6 +15,14 @@
 Tensor = torch.Tensor
 from torch import _VF
 
+# Set a global declaring that we have opt_einsum
+from importlib.util import find_spec as _find_spec
+if _find_spec('opt_einsum') is not None:
+    import opt_einsum as _opt_einsum  # type: ignore[import]
+else:
+    _opt_einsum = None
```
Comment (malfet) [inline torch/functional.py:23]: Imo `try:except:` is more pythonic way than `if :else`
```suggestion
try:
    import opt_einsum as _opt_einsum  # type: ignore[import]
except ModuleNotFoundError:
    _opt_einsum = None
```

File: .github/workflows/_mac-test.yml [hunk 2]
```diff
@@ -142,7 +142,7 @@ jobs:
           export PR_BODY="${PR_BODY//[\'\"]}"
           arch
 
-          ${CONDA_RUN} python3 -mpip install dist/*.whl
+          ${CONDA_RUN} python3 -mpip install $(echo dist/*.whl)[opt-einsum]
```
Comment (janeyx99) [inline .github/workflows/_mac-test.yml:145]: The echo is necessary because path expansion doesn't occur anymore with the addition of `[opt-einsum]`

**Review by huydhn — APPROVED** [review]
LGTM!

**Review by soulitzer — APPROVED** [review]
Cool!

**Pull request closed without merging**